added stats to indicate rt errors

bookshelfdave opened the pull request:
If a source node detects a RT error, increment the rt_source_error or rt_sink_error stat. Any increment to rt_source_error or rt_sink_error will automatically incremement rt_dirty. The rt_dirty value is written to a file to persist across restarts. When any of these stats are incremented and their value == 1, the node is registered with the FS coordinator (if it's running). Also, if a node is restarted, it will again try to register itself with the coordinator if its rt_dirty value > 0. When a fullsync is complete, the coordinator will clear rt_dirty (and remove the rt_dirty file from the filesystem) from all registered dirty nodes. If any nodes become dirty during fullsync, keep them rt_dirty even if the fullsync completes. The next pass at fullsync will clear this status. 

Note: if there is some type of disconnect between the source node and the coordinator, restarting the source node will re-register with the coordinator if it's still dirty.

misc:
you can test json stats with this:
curl -q http://localhost:8091/riak-repl/stats | python -mjson.tool

Commit:
add rt_dirty to indicate fullsync may be required

lordnull commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -628,3 +672,60 @@ is_fullsync_in_progress(State) ->
             true
     end.
+
+is_running_priv(State) ->
```
There should be a function at the bottom called is_fullsync_in_progress, which takes the state.

bookshelfdave commented:
Looking into the case where realtime is enabled and running, but fullsync is not enabled.

bookshelfdave commented on src/riak_repl_stats.erl:
```diff
@@ -98,6 +104,34 @@ elections_elected() ->
+    % then notify the server. otherwise, we don't want to spam the
+    % coordinator. Once this value is reset to 0, we'll notify the
+    % coordinator again.
+    case Stat > 0 andalso Stat < 5 of
```
hacky, I know

Commit:
removed bad test code

bookshelfdave commented:
riak_test code for this located at:
https://github.com/basho/riak_test/pull/141

lordnull commented:
eunit tests failed. If the rt dirty file reports the node is dirty, the stats set up is dependant on fscoordinator:node_dirty/1, which calls risk_core_cluster_mgr:get_leader/0 and fscooridinator_sup:started/1.

bookshelfdave commented:
@lordnull do you have the latest commit (db47f06) that "fixes" eunit issues? All 86 tests pass for me, although I have seen some of the connection mgr tests fail in the past.

The call to fscoordinator:node_dirty/1 won't do anything if the coordinator isn't running. I'm thinking of putting in some code after a fullsync completes that rpc:multicalls every node in the cluster (minus nodes that are currently marked as dirty during a fullsync) to clear the rt_dirty state. The good thing is that the node will stay marked dirty.

Edit: this change makes registering an rt_dirty node somewhat redundant because the coord will clear all nodes anyways. But I think to leave it in is fine, as rt_dirty nodes will show up in coordinator stats across the cluster.

Commit:
send a clear message to ALL nodes

This is to prevent some nodes from getting stuck in a dirty state. For
example, if the coordinator isn't running, but a rt_dirty is set, the
coordinator won't know about the dirty state. It's still useful to have
the coordinator maintain a list of dirty nodes, as they appear in the
fscoord stats.

buddhisthead commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -628,3 +672,47 @@ is_fullsync_in_progress(State) ->
+            NodesToNotify = ordsets:subtract(DirtyNodes, DirtyNodesDuringFS),
+            lager:info("Notifying nodes ~p", [ NodesToNotify]),
+            rpc:multicall(NodesToNotify, riak_repl_stats, clear_rt_dirty, []),
+            % TODO: Check return values, only clear those that passed
```
Is this TODO important to implement before we ship this next build? I presume that if the RPC call fails, then the node won't have it's dirty flag cleared and it will re-post it's status of dirty. So, it's ok if it fails?

bookshelfdave replied:
@buddhisthead no, it's not important for the release and should be removed. I think it's important that if the call fails, the node will still stay "dirty".

In the latest commit that I made, I actually rpc:multicall to all nodes (minus those in the DirtyNodesDuringFS set) to clear dirty status in case the coord wasn't running and the dirty status wasn't registered.

buddhisthead replied:
Yeah, I like that broad spectrum approach. Simplifies things and reduces code paths.

Commit:
removed leftover TODO

buddhisthead commented:
Are you planning on removing all these lager:info statements or at least changing them to lager:debug? I think that normal operation shouldn't produce so much output, even though it's handy for debugging purposes. At least we shouldn't print anything to the console when there are no dirty nodes. Silence means it's OK.

buddhisthead commented:
Works as advertised on a running cluster. I tested by starting up basho_bench and shutting down half of the six sink nodes in the other cluster. I was surprised that only two of the source nodes out of six reported errors, but I guess they were not connected uniformly to the sinks. After doing a fullsync, all dirty flags were cleared.

Nice.

If you remove/change the lager:info statements, I would happily give this a +1.

Commit:
rt_dirty lager:info -> lager:debug

Commit:
more lager:info -> lager:debug, impl node_clean

node_clean may be useful during production support

bookshelfdave commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -177,30 +207,43 @@ handle_call(status, _From, State = #state{socket=Socket}) ->
+        end,
+    {reply, ok, NewState};
+
+handle_call({node_clean, Node}, _From, State = #state{dirty_nodes=DirtyNodes}) ->
```
This may be useful if something goes wrong at runtime.

bookshelfdave commented:
@buddhisthead @lordnull changes pushed, thanks for the review!
